## Chloe's Scratchpad for the Tennis Predictions Group Project

### imports

In [1]:
# imports

import pandas as pd
import numpy as np
import regex as re

# Custom Helper Files
from prepare import *

# Split 
from sklearn.model_selection import train_test_split

# Stats
from scipy import stats

# Model
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import export_graphviz
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression

# Visualize
import matplotlib.pyplot as plt
import seaborn as sns

# Ignore Warnings
import warnings
warnings.filterwarnings("ignore")

# Remove Limits On Viewing Dataframes
pd.set_option('display.max_columns', None)

In [2]:
# read .csv
# df = pd.read_csv('PlayerData.csv')

In [3]:
# df.head()

In [4]:
# # read .csv
# player1 = pd.read_csv('ATPUpdated.csv')

In [5]:
# player1.head()

In [6]:
# player1.columns

In [7]:
# player2 = player1.copy(0)

In [8]:
# player2.head()

In [9]:
# player2 = player2.rename(columns = {'Player1_id': 'Player2_id', 'Player1_seed': 'Player2_seed', 'Player1_entry': 'Player2_entry', 'Player1_name': 'Player2_name', 'Player1_hand': 'Player2_hand', 'Player1_ht': 'Player2_ht', 'Player1_ioc': 'Player2_ioc', 'Player1_age': 'Player2_age', 'Player2_id': 'Player1_id', 'Player2_seed': 'Player1_seed', 'Player2_entry': 'Player1_entry', 'Player2_name': 'Player1_name', 'Player2_hand': 'Player1_hand', 'Player2_ht': 'Player1_ht', 'Player2_ioc': 'Player1_ioc', 'Player2_age': 'Player1_age'})
                                       
                                       

In [10]:
# player2.head()

In [11]:
# player2 = player2.rename(columns = {'Player1_aces': 'Player2_aces', 'Player1_double_faults': 'Player2_double_faults', 'Player1_service_points': 'Player2_service_points', 'Player1_first_serves_in': 'Player2_first_serves_in', 'Player1_first_serve_points_won': 'Player2_first_serve_points_won', 'Player1_second_serve_points_won': 'Player2_second_serve_points_won', 'Player1_service_game_total': 'Player2_service_game_total', 'Player1_break_points_saved': 'Player2_break_points_saved', 'Player1_break_points_faced': 'Player2_break_points_faced', 'Player2_aces': 'Player1_aces', 'Player2_double_faults': 'Player1_double_faults', 'Player2_service_points': 'Player1_service_points', 'Player2_first_serves_in': 'Player1_first_serves_in', 'Player2_first_serve_points_won': 'Player1_first_serve_points_won', 'Player2_second_serve_points_won': 'Player1_second_serve_points_won', 'Player2_service_game_total': 'Player1_service_game_total', 'Player2_break_points_saved': 'Player1_break_points_saved', 'Player2_break_points_faced': 'Player1_break_points_faced'})



In [12]:
# player2.head()

In [13]:
# player2 = player2.rename(columns = {'Player1_rank': 'Player2_rank', 'Player1_rank_points': 'Player2_rank_points', 'Player2_rank': 'Player1_rank', 'Player2_rank_points': 'Player1_rank_points', 'Player1_first_serve_%': 'Player2_first_serve_%', 'Player1_first_serve_win_%': 'Player2_first_serve_win_%', 'Player1_break_points_won': 'Player2_break_points_won', 'Player2_first_serve_%': 'Player1_first_serve_%', 'Player2_first_serve_win_%': 'Player1_first_serve_win_%', 'Player2_break_points_won': 'Player1_break_points_won'})



In [14]:
# player2.head()

In [15]:
# # read .csv
# df1 = pd.read_csv('Player1.csv')

In [16]:
# df1.head(1)

In [17]:
# # read .csv
# df2 = pd.read_csv('Player2.csv')

In [18]:
# df2.head(1)

Concat the two frames together

In [19]:
# frames = [player1, player2]

In [20]:
# combined_df = pd.concat(frames)

In [21]:
# combined_df.head(1)

In [22]:
# combined_df.shape

In [23]:
# combined_df.sample(10)

In [24]:
# combined_df.match_num.unique().shape

In [25]:
# # save df to csv in current location
# combined_df.to_csv('ATPUpdated_Combined.csv')

In [26]:
# # read .csv
# df = pd.read_csv('ATPUpdated_Combined.csv')

In [27]:
# df.head()

In [28]:
# df.shape

In [29]:
# # check to see if combined properly
# df[df.Player1_name.isin(['Kei Nishikori'])]

In [30]:
# # check to see if combined properly
# df[df.Player2_name.isin(['Kei Nishikori'])]

### Acquire 

Changed winner/loser to player1/player2 based on whose name came alphabetically first (outsourced code for that part). 

Dataset with randomized player1 and player2: 

In [31]:
# # read .csv
# df = pd.read_csv('ATPMain.csv')

In [32]:
# df = prepare_atp()

In [33]:
# df.head()

### Prepare

In [34]:
# # Drop duplicates if any

# print(df.shape)
# df = df.drop_duplicates()
# print(df.shape)

There are no duplicates.

In [35]:
# df.isnull().sum()

My goal is modeling (MVP) so I'm going to add only the columns that i need to a new copied df, and prepare it from there. 

In [36]:
# df.columns

In [37]:
# dfm = df[['tourney_id', 'surface', 'draw_size',
#        'tourney_level', 'best_of',
#        'round', 'player_1', 'player_2', 'player_1_age',
#        'player_2_age', 'player_1_hand',
#        'player_2_hand', 'player_1_ht', 'player_2_ht', 'player_1_ioc', 'player_2_ioc', 'player_1_name',
#        'player_2_name', 'player_1_rank', 'player_2_rank',
#        'player_1_rank_points', 'player_2_rank_points', 'winner']].copy(0)

In [38]:
# dfm.head()

In [39]:
# dfm.surface.describe()

In [40]:
# # create dummy columns for surface column and concat to df
# dummy_df = pd.get_dummies(dfm[['surface', 'tourney_level', 'player_1_hand', 'player_2_hand', 'round']], dummy_na=False, drop_first=[False])
# dfm = pd.concat([dfm, dummy_df], axis=1)

In [41]:
# dfm.head()

In [42]:
# dfm.dtypes

In [43]:
# dfm['player_1_wins'] = np.where(dfm['winner'] == dfm['player_1_name'], True, False)
# dfm.head()

In [44]:
# dfm['player_1_wins'] = np.where(dfm['winner'] == dfm['player_1_name'], 1, 0)
# dfm.head()

In [45]:
# dfm.dtypes

In [46]:
# dfm.head()

In [47]:
# dfm.dtypes

In [48]:
# def prepare_atp():

#     # read .csv into a dataframe
#     df = pd.read_csv('ATPMain.csv', index_col = 0)

#     # set index to the tournament start date
#     df = df.set_index('tourney_date')

#     # convert index to datetime object and sort in ascending order
#     df.index = pd.to_datetime(df.index, format = '%Y%m%d')
#     df = df.sort_index(ascending = True)

#     # create target variable 
#     df['player_1_wins'] = np.where(df['winner'] == df['player_1_name'], True, False)

#     # create dummy columns for surface, level, hand, and round 
#     dummy_df = pd.get_dummies(df[['surface', 'tourney_level', 'player_1_hand', 'player_2_hand', 'round']], dummy_na=False, drop_first=[False])
#     # concat dummy columns to df
#     df = pd.concat([df, dummy_df], axis=1)
    

#     # drop all walkovers (no useful stats) and best of 1 matches (extremely rare format)
#     df = df.drop(df[df.score == 'W/O'].index)
#     df = df.drop(df[df.best_of == 1].index)

#     df_clean = df = df.dropna(subset=['player_1_aces'])
    
#     return df

In [49]:
df = prepare_atp()

In [50]:
df.head()

tourney_id tourney_name surface  draw_size tourney_level  \
tourney_date                                                            
1999-01-11     1999-338       Sydney    Hard         32             A   
1999-01-11     1999-338       Sydney    Hard         32             A   
1999-01-11     1999-338       Sydney    Hard         32             A   
1999-01-11     1999-338       Sydney    Hard         32             A   
1999-01-11     1999-338       Sydney    Hard         32             A   

              match_num        score  best_of round  minutes        player_1  \
tourney_date                                                                   
1999-01-11           16   7-6(1) 6-1        3   R32     84.0  Lleyton Hewitt   
1999-01-11           15  5-7 6-3 6-2        3   R32    115.0  Martin Damm Sr   
1999-01-11           14      6-2 6-4        3   R32     61.0   Jan Siemerink   
1999-01-11           13      6-1 6-3        3   R32     43.0    Hicham Arazi   
1999-01-11           12   6-3 7-6(4)        3   R32     85.0     Carlos Moya   

                      player_2  player_1_age  player_2_age player_1_entry  \
tourney_date                                                                
1999-01-11      Patrick Rafter     17.878166     26.036961            NaN   
1999-01-11      Nicolas Kiefer     26.444901     21.519507              Q   
1999-01-11      Mariano Puerta     28.744695     20.312115            NaN   
1999-01-11         Todd Martin     25.229295     28.511978            NaN   
1999-01-11    Thomas Johansson     22.373717     23.802875            NaN   

             player_2_entry player_1_hand player_2_hand  player_1_ht  \
tourney_date                                                           
1999-01-11              NaN             R             R        180.0   
1999-01-11              NaN             R             R        188.0   
1999-01-11              NaN             L             L        183.0   
1999-01-11              NaN             L             R        175.0   
1999-01-11              NaN             R             R        190.0   

              player_2_ht  player_1_id  player_2_id player_1_ioc player_2_ioc  \
tourney_date                                                                    
1999-01-11          185.0       103720       102158          AUS          AUS   
1999-01-11          183.0       210013       103017          CZE          GER   
1999-01-11          180.0       101733       103264          NED          ARG   
1999-01-11          198.0       102271       101774          MAR          USA   
1999-01-11          180.0       102845       102563          ESP          SWE   

               player_1_name     player_2_name  player_1_rank  player_2_rank  \
tourney_date                                                                   
1999-01-11    Lleyton Hewitt    Patrick Rafter          104.0            4.0   
1999-01-11    Martin Damm Sr    Nicolas Kiefer           75.0           36.0   
1999-01-11     Jan Siemerink    Mariano Puerta           19.0           38.0   
1999-01-11      Hicham Arazi       Todd Martin           34.0           16.0   
1999-01-11       Carlos Moya  Thomas Johansson            5.0           17.0   

              player_1_rank_points  player_2_rank_points  player_1_seed  \
tourney_date                                                              
1999-01-11                   456.0                3315.0            NaN   
1999-01-11                   657.0                1007.0            NaN   
1999-01-11                  1664.0                 983.0            NaN   
1999-01-11                  1069.0                1774.0            NaN   
1999-01-11                  3159.0                1761.0            3.0   

              player_2_seed  player_1_aces  player_2_aces  \
tourney_date                                                
1999-01-11              2.0            2.0            0.0   
1999-01-11              NaN            8.0            6.0

In [51]:
def prepare_atp():

    # read .csv into a dataframe
    df = pd.read_csv('ATPMain.csv', index_col = 0)

    # set index to the tournament start date
    df = df.set_index('tourney_date')

    # convert index to datetime object and sort in ascending order
    df.index = pd.to_datetime(df.index, format = '%Y%m%d')
    df = df.sort_index(ascending = True)

    # set dataframe to years 1999-2019 (21 years of recent data minus corona years)
    df = df['1999-01-01':'2020-01-01']

    # create target variable 
    df['player_1_wins'] = np.where(df['winner'] == df['player_1_name'], True, False)
    
    # drop all walkovers (no useful stats) and best of 1 matches (extremely rare format)
    df = df.drop(df[df.score == 'W/O'].index)
    df = df.drop(df[df.best_of == 1].index)

    # drop janky records
    df = df[df['player_1_first_serve_%'].notnull()]
    df = df[df['player_2_first_serve_%'].notnull()]

    # gather all players in two sets
    players = set(list(df.player_1.unique()))
    players2 = set(list(df.player_2.unique()))

    # combine, drop duplicates, convert set to list type and sort for fun 
    players = players.union(players2)
    players = list(players)
    players.sort()

    # set an empty list to fill
    less_than_50 = []

    # use a for-loop to run through the list of players
    for player in players:
        # set up if-conditional to see if the length of records where a player shows up in the dataframe is less than 50
        if len(df[(df.player_1 == player) | (df.player_2 == player)]) < 50:
            # if there are less than 50 records, add the player's name to the list
            less_than_50.append(player)
    # sort list
    less_than_50.sort()

    # run through loop of list of players with less than 50 matches
    for player in less_than_50:
        # set dataframe to records where these players are not present
        df = df[df.player_1 != player]
        df = df[df.player_2 != player]

    df_clean = df = df.dropna(subset=['player_1_aces'])

    # create dummy columns for surface, level, hand, and round 
    dummy_df = pd.get_dummies(df[['surface', 'tourney_level', 'player_1_hand', 'player_2_hand', 'round']], dummy_na = False, drop_first = False)
    # concat dummy columns to df
    df = pd.concat([df, dummy_df], axis=1)
    
    return df

In [52]:
df = prepare_atp()

In [53]:
df.head(1)

tourney_id tourney_name surface  draw_size tourney_level  \
tourney_date                                                            
1999-01-11     1999-338       Sydney    Hard         32             A   

              match_num       score  best_of round  minutes        player_1  \
tourney_date                                                                  
1999-01-11           16  7-6(1) 6-1        3   R32     84.0  Lleyton Hewitt   

                    player_2  player_1_age  player_2_age player_1_entry  \
tourney_date                                                              
1999-01-11    Patrick Rafter     17.878166     26.036961            NaN   

             player_2_entry player_1_hand player_2_hand  player_1_ht  \
tourney_date                                                           
1999-01-11              NaN             R             R        180.0   

              player_2_ht  player_1_id  player_2_id player_1_ioc player_2_ioc  \
tourney_date                                                                    
1999-01-11          185.0       103720       102158          AUS          AUS   

               player_1_name   player_2_name  player_1_rank  player_2_rank  \
tourney_date                                                                 
1999-01-11    Lleyton Hewitt  Patrick Rafter          104.0            4.0   

              player_1_rank_points  player_2_rank_points  player_1_seed  \
tourney_date                                                              
1999-01-11                   456.0                3315.0            NaN   

              player_2_seed  player_1_aces  player_2_aces  \
tourney_date                                                
1999-01-11              2.0            2.0            0.0   

              player_1_double_faults  player_2_double_faults  \
tourney_date                                                   
1999-01-11                       0.0                     5.0   

              player_1_service_points  player_2_service_points  \
tourney_date                                                     
1999-01-11                       73.0                     59.0   

              player_1_first_serves_in  player_2_first_serves_in  \
tourney_date                                                       
1999-01-11                        51.0                      36.0   

              player_1_first_serve_points_won  \
tourney_date                                    
1999-01-11                               32.0   

              player_2_first_serve_points_won  \
tourney_date                                    
1999-01-11                               24.0   

              player_1_second_serve_points_won  \
tourney_date                                     
1999-01-11                                14.0   

              player_2_second_serve_points_won  player_1_service_game_total  \
tourney_date                                                                  
1999-01-11                                 8.0                         10.0   

              player_2_service_game_total  player_1_break_points_saved  \
tourney_date                                                             
1999-01-11                            9.0                          4.0   

              player_2_break_points_saved  player_1_break_points_faced  \
tourney_date                                                             
1999-01-11                            3.0                          6.0   

              player_2_break_points_faced          winner  \
tourney_date                                                
1999-01-11                            7.0  Lleyton Hewitt   

              player_1_first_serve_%  player_2_first_serve_%  \
tourney_date                                                   
1999-01-11                   0.69863                0.610169   

              player_1_first_serve_win_%  player_2_first_serve_win_%  \
tourney_date                                  

Create a column with the total games played in each match. 

In [54]:
# function to create a column adding all of the games in the score column together using regex

def total_games(s):
    num_list = re.findall('\d+', s)
    num_list = [int(x) for x in num_list]
    return sum(num_list)

df['total_games'] = df['score'].apply(total_games)
df

tourney_id  \
tourney_date                                    
1999-01-11                           1999-338   
1999-01-11                           1999-338   
1999-01-11                           1999-338   
1999-01-11                           1999-338   
1999-01-11                           1999-338   
...                                       ...   
2019-11-23    2019-M-DC-2019-FLS-M-GBR-ESP-01   
2019-11-23    2019-M-DC-2019-FLS-M-RUS-CAN-01   
2019-11-23    2019-M-DC-2019-FLS-M-RUS-CAN-01   
2019-11-23    2019-M-DC-2019-FLS-M-GBR-ESP-01   
2019-11-24    2019-M-DC-2019-FLS-M-CAN-ESP-01   

                                 tourney_name surface  draw_size  \
tourney_date                                                       
1999-01-11                             Sydney    Hard         32   
1999-01-11                             Sydney    Hard         32   
1999-01-11                             Sydney    Hard         32   
1999-01-11                             Sydney    Hard         32   
1999-01-11                             Sydney    Hard         32   
...                                       ...     ...        ...   
2019-11-23    Davis Cup Finals SF: GBR vs ESP    Hard          4   
2019-11-23    Davis Cup Finals SF: RUS vs CAN    Hard          4   
2019-11-23    Davis Cup Finals SF: RUS vs CAN    Hard          4   
2019-11-23    Davis Cup Finals SF: GBR vs ESP    Hard          4   
2019-11-24     Davis Cup Finals F: CAN vs ESP    Hard          4   

             tourney_level  match_num        score  best_of round  minutes  \
tourney_date                                                                 
1999-01-11               A         16   7-6(1) 6-1        3   R32     84.0   
1999-01-11               A         15  5-7 6-3 6-2        3   R32    115.0   
1999-01-11               A         14      6-2 6-4        3   R32     61.0   
1999-01-11               A         13      6-1 6-3        3   R32     43.0   
1999-01-11               A         12   6-3 7-6(4)        3   R32     85.0   
...                    ...        ...          ...      ...   ...      ...   
2019-11-23               D          2      6-4 6-0        3    RR     85.0   
2019-11-23               D          1      6-4 6-4        3    RR     85.0   
2019-11-23               D          2  6-4 4-6 6-4        3    RR    110.0   
2019-11-23               D          1   6-3 7-6(3)        3    RR     81.0   
2019-11-24               D          2   6-3 7-6(7)        3    RR    115.0   

                      player_1          player_2  player_1_age  player_2_age  \
tourney_date                                                                   
1999-01-11      Lleyton Hewitt    Patrick Rafter     17.878166     26.036961   
1999-01-11      Martin Damm Sr    Nicolas Kiefer     26.444901     21.519507   
1999-01-11       Jan Siemerink    Mariano Puerta     28.744695     20.312115   
1999-01-11        Hicham Arazi       Todd Martin     25.229295     28.511978   
1999-01-11         Carlos Moya  Thomas Johansson     22.373717     23.802875   
...                        ...               ...           ...           ...   
2019-11-23        Daniel Evans      Rafael Nadal     29.489391     33.459274   
2019-11-23       Andrey Rublev    Vasek Pospisil     22.078029     29.404517   
2019-11-23    Denis Shapovalov   Karen Khachanov     20.594114     23.493498   
2019-11-23     Feliciano Lopez       Kyle Edmund     38.160164     24.859685   
2019-11-24    Denis Shapovalov      Rafael Nadal     20.594114     33.459274   

             player_1_entry player_2_entry player_1_hand player_2_hand  \
tourney_date                                                             
1999-01-11              NaN            NaN             R             R   
1999-01-11                Q            NaN             R             R   
1999-01-11              NaN            NaN             L             L   
1999-01-11              NaN            NaN             L             R   
1999-01-11  

In [55]:
df.shape

(35969, 81)

In [56]:
# removes rows where the score contains letters that indicate a retirement, default, or walkover

score_mask = df['score'].apply(lambda x: bool(re.findall('[a-zA-Z]', x)))
df = df[~score_mask]
df.head(1)

tourney_id tourney_name surface  draw_size tourney_level  \
tourney_date                                                            
1999-01-11     1999-338       Sydney    Hard         32             A   

              match_num       score  best_of round  minutes        player_1  \
tourney_date                                                                  
1999-01-11           16  7-6(1) 6-1        3   R32     84.0  Lleyton Hewitt   

                    player_2  player_1_age  player_2_age player_1_entry  \
tourney_date                                                              
1999-01-11    Patrick Rafter     17.878166     26.036961            NaN   

             player_2_entry player_1_hand player_2_hand  player_1_ht  \
tourney_date                                                           
1999-01-11              NaN             R             R        180.0   

              player_2_ht  player_1_id  player_2_id player_1_ioc player_2_ioc  \
tourney_date                                                                    
1999-01-11          185.0       103720       102158          AUS          AUS   

               player_1_name   player_2_name  player_1_rank  player_2_rank  \
tourney_date                                                                 
1999-01-11    Lleyton Hewitt  Patrick Rafter          104.0            4.0   

              player_1_rank_points  player_2_rank_points  player_1_seed  \
tourney_date                                                              
1999-01-11                   456.0                3315.0            NaN   

              player_2_seed  player_1_aces  player_2_aces  \
tourney_date                                                
1999-01-11              2.0            2.0            0.0   

              player_1_double_faults  player_2_double_faults  \
tourney_date                                                   
1999-01-11                       0.0                     5.0   

              player_1_service_points  player_2_service_points  \
tourney_date                                                     
1999-01-11                       73.0                     59.0   

              player_1_first_serves_in  player_2_first_serves_in  \
tourney_date                                                       
1999-01-11                        51.0                      36.0   

              player_1_first_serve_points_won  \
tourney_date                                    
1999-01-11                               32.0   

              player_2_first_serve_points_won  \
tourney_date                                    
1999-01-11                               24.0   

              player_1_second_serve_points_won  \
tourney_date                                     
1999-01-11                                14.0   

              player_2_second_serve_points_won  player_1_service_game_total  \
tourney_date                                                                  
1999-01-11                                 8.0                         10.0   

              player_2_service_game_total  player_1_break_points_saved  \
tourney_date                                                             
1999-01-11                            9.0                          4.0   

              player_2_break_points_saved  player_1_break_points_faced  \
tourney_date                                                             
1999-01-11                            3.0                          6.0   

              player_2_break_points_faced          winner  \
tourney_date                                                
1999-01-11                            7.0  Lleyton Hewitt   

              player_1_first_serve_%  player_2_first_serve_%  \
tourney_date                                                   
1999-01-11                   0.69863                0.610169   

              player_1_first_serve_win_%  player_2_first_serve_win_%  \
tourney_date                                  

In [57]:
df.shape

(34791, 81)

In [58]:
# Remove Limits On Viewing Dataframes
pd.set_option('display.max_rows', None)

In [59]:
df.isnull().sum()

tourney_id                              0
tourney_name                            0
surface                                 0
draw_size                               0
tourney_level                           0
match_num                               0
score                                   0
best_of                                 0
round                                   0
minutes                               833
player_1                                0
player_2                                0
player_1_age                            0
player_2_age                            0
player_1_entry                      30322
player_2_entry                      30185
player_1_hand                           0
player_2_hand                           0
player_1_ht                            84
player_2_ht                            69
player_1_id                             0
player_2_id                             0
player_1_ioc                            0
player_2_ioc                      

In [60]:
dfm = df[['draw_size', 'winner',
       'tourney_level', 'best_of', 'player_1', 'player_2', 'player_1_age',
       'player_2_age', 'player_1_hand',
       'player_2_hand', 'player_1_ht', 'player_2_ht', 'player_1_ioc', 'player_2_ioc', 'player_1_name',
       'player_2_name', 'player_1_rank', 'player_2_rank',
       'player_1_rank_points', 'player_2_rank_points', 'player_1_wins', 'round_ER', 'round_F', 'round_QF', 'round_R128', 'round_R16', 'round_R32', 'round_R64', 'round_RR', 'round_SF', 'player_1_hand_R', 'player_1_hand_L', 
       'tourney_level_A', 'tourney_level_D', 'tourney_level_F', 'tourney_level_G', 'tourney_level_M', 'surface_Carpet', 'surface_Clay',
       'surface_Grass', 'surface_Hard']].copy(0)

In [61]:
dfm.head()

draw_size          winner tourney_level  best_of  \
tourney_date                                                     
1999-01-11           32  Lleyton Hewitt             A        3   
1999-01-11           32  Nicolas Kiefer             A        3   
1999-01-11           32  Mariano Puerta             A        3   
1999-01-11           32     Todd Martin             A        3   
1999-01-11           32     Carlos Moya             A        3   

                    player_1          player_2  player_1_age  player_2_age  \
tourney_date                                                                 
1999-01-11    Lleyton Hewitt    Patrick Rafter     17.878166     26.036961   
1999-01-11    Martin Damm Sr    Nicolas Kiefer     26.444901     21.519507   
1999-01-11     Jan Siemerink    Mariano Puerta     28.744695     20.312115   
1999-01-11      Hicham Arazi       Todd Martin     25.229295     28.511978   
1999-01-11       Carlos Moya  Thomas Johansson     22.373717     23.802875   

             player_1_hand player_2_hand  player_1_ht  player_2_ht  \
tourney_date                                                         
1999-01-11               R             R        180.0        185.0   
1999-01-11               R             R        188.0        183.0   
1999-01-11               L             L        183.0        180.0   
1999-01-11               L             R        175.0        198.0   
1999-01-11               R             R        190.0        180.0   

             player_1_ioc player_2_ioc   player_1_name     player_2_name  \
tourney_date                                                               
1999-01-11            AUS          AUS  Lleyton Hewitt    Patrick Rafter   
1999-01-11            CZE          GER  Martin Damm Sr    Nicolas Kiefer   
1999-01-11            NED          ARG   Jan Siemerink    Mariano Puerta   
1999-01-11            MAR          USA    Hicham Arazi       Todd Martin   
1999-01-11            ESP          SWE     Carlos Moya  Thomas Johansson   

              player_1_rank  player_2_rank  player_1_rank_points  \
tourney_date                                                       
1999-01-11            104.0            4.0                 456.0   
1999-01-11             75.0           36.0                 657.0   
1999-01-11             19.0           38.0                1664.0   
1999-01-11             34.0           16.0                1069.0   
1999-01-11              5.0           17.0                3159.0   

              player_2_rank_points  player_1_wins  round_ER  round_F  \
tourney_date                                                           
1999-01-11                  3315.0           True         0        0   
1999-01-11                  1007.0          False         0        0   
1999-01-11                   983.0          False         0        0   
1999-01-11                  1774.0          False         0        0   
1999-01-11                  1761.0           True         0        0   

              round_QF  round_R128  round_R16  round_R32  round_R64  round_RR  \
tourney_date                                                                    
1999-01-11           0           0          0          1          0         0   
1999-01-11           0           0          0          1          0         0   
1999-01-11           0           0          0          1          0         0   
1999-01-11           0           0          0          1          0         0   
1999-01-11           0           0          0          1          0         0   

              round_SF  player_1_hand_R  player_1_hand_L  tourney_level_A  \
tourney_date                                                                
1999-01-11           0                1                0                1   
1999-01-11           0                1                0                1   
1999-01-11           0                0                1                1   
1999-01-11           0                0      

In [62]:
dfm.isnull().sum()

draw_size                0
winner                   0
tourney_level            0
best_of                  0
player_1                 0
player_2                 0
player_1_age             0
player_2_age             0
player_1_hand            0
player_2_hand            0
player_1_ht             84
player_2_ht             69
player_1_ioc             0
player_2_ioc             0
player_1_name            0
player_2_name            0
player_1_rank           16
player_2_rank           15
player_1_rank_points    16
player_2_rank_points    15
player_1_wins            0
round_ER                 0
round_F                  0
round_QF                 0
round_R128               0
round_R16                0
round_R32                0
round_R64                0
round_RR                 0
round_SF                 0
player_1_hand_R          0
player_1_hand_L          0
tourney_level_A          0
tourney_level_D          0
tourney_level_F          0
tourney_level_G          0
tourney_level_M          0
s

In [64]:
# drop null rows in specific columns
dfm = dfm[dfm.player_1_hand.notnull()]
dfm = dfm[dfm.player_2_hand.notnull()]
dfm = dfm[dfm.player_1_ht.notnull()]
dfm = dfm[dfm.player_2_ht.notnull()]
dfm = dfm[dfm.player_1_rank.notnull()]
dfm = dfm[dfm.player_2_rank.notnull()]
dfm = dfm[dfm.player_1_rank_points.notnull()]
dfm = dfm[dfm.player_2_rank_points.notnull()]

In [65]:
dfm.isnull().sum()

draw_size               0
winner                  0
tourney_level           0
best_of                 0
player_1                0
player_2                0
player_1_age            0
player_2_age            0
player_1_hand           0
player_2_hand           0
player_1_ht             0
player_2_ht             0
player_1_ioc            0
player_2_ioc            0
player_1_name           0
player_2_name           0
player_1_rank           0
player_2_rank           0
player_1_rank_points    0
player_2_rank_points    0
player_1_wins           0
round_ER                0
round_F                 0
round_QF                0
round_R128              0
round_R16               0
round_R32               0
round_R64               0
round_RR                0
round_SF                0
player_1_hand_R         0
player_1_hand_L         0
tourney_level_A         0
tourney_level_D         0
tourney_level_F         0
tourney_level_G         0
tourney_level_M         0
surface_Carpet          0
surface_Clay

In [66]:
dfm['winner_rank'] = np.where(dfm['winner'] == dfm['player_1_name'], dfm['player_1_rank'], dfm['player_2_rank'])
dfm.head()

draw_size          winner tourney_level  best_of  \
tourney_date                                                     
1999-01-11           32  Lleyton Hewitt             A        3   
1999-01-11           32  Nicolas Kiefer             A        3   
1999-01-11           32  Mariano Puerta             A        3   
1999-01-11           32     Todd Martin             A        3   
1999-01-11           32     Carlos Moya             A        3   

                    player_1          player_2  player_1_age  player_2_age  \
tourney_date                                                                 
1999-01-11    Lleyton Hewitt    Patrick Rafter     17.878166     26.036961   
1999-01-11    Martin Damm Sr    Nicolas Kiefer     26.444901     21.519507   
1999-01-11     Jan Siemerink    Mariano Puerta     28.744695     20.312115   
1999-01-11      Hicham Arazi       Todd Martin     25.229295     28.511978   
1999-01-11       Carlos Moya  Thomas Johansson     22.373717     23.802875   

             player_1_hand player_2_hand  player_1_ht  player_2_ht  \
tourney_date                                                         
1999-01-11               R             R        180.0        185.0   
1999-01-11               R             R        188.0        183.0   
1999-01-11               L             L        183.0        180.0   
1999-01-11               L             R        175.0        198.0   
1999-01-11               R             R        190.0        180.0   

             player_1_ioc player_2_ioc   player_1_name     player_2_name  \
tourney_date                                                               
1999-01-11            AUS          AUS  Lleyton Hewitt    Patrick Rafter   
1999-01-11            CZE          GER  Martin Damm Sr    Nicolas Kiefer   
1999-01-11            NED          ARG   Jan Siemerink    Mariano Puerta   
1999-01-11            MAR          USA    Hicham Arazi       Todd Martin   
1999-01-11            ESP          SWE     Carlos Moya  Thomas Johansson   

              player_1_rank  player_2_rank  player_1_rank_points  \
tourney_date                                                       
1999-01-11            104.0            4.0                 456.0   
1999-01-11             75.0           36.0                 657.0   
1999-01-11             19.0           38.0                1664.0   
1999-01-11             34.0           16.0                1069.0   
1999-01-11              5.0           17.0                3159.0   

              player_2_rank_points  player_1_wins  round_ER  round_F  \
tourney_date                                                           
1999-01-11                  3315.0           True         0        0   
1999-01-11                  1007.0          False         0        0   
1999-01-11                   983.0          False         0        0   
1999-01-11                  1774.0          False         0        0   
1999-01-11                  1761.0           True         0        0   

              round_QF  round_R128  round_R16  round_R32  round_R64  round_RR  \
tourney_date                                                                    
1999-01-11           0           0          0          1          0         0   
1999-01-11           0           0          0          1          0         0   
1999-01-11           0           0          0          1          0         0   
1999-01-11           0           0          0          1          0         0   
1999-01-11           0           0          0          1          0         0   

              round_SF  player_1_hand_R  player_1_hand_L  tourney_level_A  \
tourney_date                                                                
1999-01-11           0                1                0                1   
1999-01-11           0                1                0                1   
1999-01-11           0                0                1                1   
1999-01-11           0                0      

In [68]:
dfm['loser_rank'] = np.where(dfm['winner'] == dfm['player_2_name'], dfm['player_1_rank'], dfm['player_2_rank'])
dfm.head()

draw_size          winner tourney_level  best_of  \
tourney_date                                                     
1999-01-11           32  Lleyton Hewitt             A        3   
1999-01-11           32  Nicolas Kiefer             A        3   
1999-01-11           32  Mariano Puerta             A        3   
1999-01-11           32     Todd Martin             A        3   
1999-01-11           32     Carlos Moya             A        3   

                    player_1          player_2  player_1_age  player_2_age  \
tourney_date                                                                 
1999-01-11    Lleyton Hewitt    Patrick Rafter     17.878166     26.036961   
1999-01-11    Martin Damm Sr    Nicolas Kiefer     26.444901     21.519507   
1999-01-11     Jan Siemerink    Mariano Puerta     28.744695     20.312115   
1999-01-11      Hicham Arazi       Todd Martin     25.229295     28.511978   
1999-01-11       Carlos Moya  Thomas Johansson     22.373717     23.802875   

             player_1_hand player_2_hand  player_1_ht  player_2_ht  \
tourney_date                                                         
1999-01-11               R             R        180.0        185.0   
1999-01-11               R             R        188.0        183.0   
1999-01-11               L             L        183.0        180.0   
1999-01-11               L             R        175.0        198.0   
1999-01-11               R             R        190.0        180.0   

             player_1_ioc player_2_ioc   player_1_name     player_2_name  \
tourney_date                                                               
1999-01-11            AUS          AUS  Lleyton Hewitt    Patrick Rafter   
1999-01-11            CZE          GER  Martin Damm Sr    Nicolas Kiefer   
1999-01-11            NED          ARG   Jan Siemerink    Mariano Puerta   
1999-01-11            MAR          USA    Hicham Arazi       Todd Martin   
1999-01-11            ESP          SWE     Carlos Moya  Thomas Johansson   

              player_1_rank  player_2_rank  player_1_rank_points  \
tourney_date                                                       
1999-01-11            104.0            4.0                 456.0   
1999-01-11             75.0           36.0                 657.0   
1999-01-11             19.0           38.0                1664.0   
1999-01-11             34.0           16.0                1069.0   
1999-01-11              5.0           17.0                3159.0   

              player_2_rank_points  player_1_wins  round_ER  round_F  \
tourney_date                                                           
1999-01-11                  3315.0           True         0        0   
1999-01-11                  1007.0          False         0        0   
1999-01-11                   983.0          False         0        0   
1999-01-11                  1774.0          False         0        0   
1999-01-11                  1761.0           True         0        0   

              round_QF  round_R128  round_R16  round_R32  round_R64  round_RR  \
tourney_date                                                                    
1999-01-11           0           0          0          1          0         0   
1999-01-11           0           0          0          1          0         0   
1999-01-11           0           0          0          1          0         0   
1999-01-11           0           0          0          1          0         0   
1999-01-11           0           0          0          1          0         0   

              round_SF  player_1_hand_R  player_1_hand_L  tourney_level_A  \
tourney_date                                                                
1999-01-11           0                1                0                1   
1999-01-11           0                1                0                1   
1999-01-11           0                0                1                1   
1999-01-11           0                0      

In [69]:
# Calculate the difference in stats between player1 and playeer2. Save to new column. 
dfm['ht_diff'] = dfm.player_1_ht - dfm.player_2_ht
dfm['age_diff'] = dfm.player_1_age - dfm.player_2_age
dfm['rank_diff'] = dfm.player_1_rank - dfm.player_2_rank
dfm['rank_points_diff'] = dfm.player_1_rank_points - dfm.player_2_rank_points

In [70]:
dfm.head()

draw_size          winner tourney_level  best_of  \
tourney_date                                                     
1999-01-11           32  Lleyton Hewitt             A        3   
1999-01-11           32  Nicolas Kiefer             A        3   
1999-01-11           32  Mariano Puerta             A        3   
1999-01-11           32     Todd Martin             A        3   
1999-01-11           32     Carlos Moya             A        3   

                    player_1          player_2  player_1_age  player_2_age  \
tourney_date                                                                 
1999-01-11    Lleyton Hewitt    Patrick Rafter     17.878166     26.036961   
1999-01-11    Martin Damm Sr    Nicolas Kiefer     26.444901     21.519507   
1999-01-11     Jan Siemerink    Mariano Puerta     28.744695     20.312115   
1999-01-11      Hicham Arazi       Todd Martin     25.229295     28.511978   
1999-01-11       Carlos Moya  Thomas Johansson     22.373717     23.802875   

             player_1_hand player_2_hand  player_1_ht  player_2_ht  \
tourney_date                                                         
1999-01-11               R             R        180.0        185.0   
1999-01-11               R             R        188.0        183.0   
1999-01-11               L             L        183.0        180.0   
1999-01-11               L             R        175.0        198.0   
1999-01-11               R             R        190.0        180.0   

             player_1_ioc player_2_ioc   player_1_name     player_2_name  \
tourney_date                                                               
1999-01-11            AUS          AUS  Lleyton Hewitt    Patrick Rafter   
1999-01-11            CZE          GER  Martin Damm Sr    Nicolas Kiefer   
1999-01-11            NED          ARG   Jan Siemerink    Mariano Puerta   
1999-01-11            MAR          USA    Hicham Arazi       Todd Martin   
1999-01-11            ESP          SWE     Carlos Moya  Thomas Johansson   

              player_1_rank  player_2_rank  player_1_rank_points  \
tourney_date                                                       
1999-01-11            104.0            4.0                 456.0   
1999-01-11             75.0           36.0                 657.0   
1999-01-11             19.0           38.0                1664.0   
1999-01-11             34.0           16.0                1069.0   
1999-01-11              5.0           17.0                3159.0   

              player_2_rank_points  player_1_wins  round_ER  round_F  \
tourney_date                                                           
1999-01-11                  3315.0           True         0        0   
1999-01-11                  1007.0          False         0        0   
1999-01-11                   983.0          False         0        0   
1999-01-11                  1774.0          False         0        0   
1999-01-11                  1761.0           True         0        0   

              round_QF  round_R128  round_R16  round_R32  round_R64  round_RR  \
tourney_date                                                                    
1999-01-11           0           0          0          1          0         0   
1999-01-11           0           0          0          1          0         0   
1999-01-11           0           0          0          1          0         0   
1999-01-11           0           0          0          1          0         0   
1999-01-11           0           0          0          1          0         0   

              round_SF  player_1_hand_R  player_1_hand_L  tourney_level_A  \
tourney_date                                                                
1999-01-11           0                1                0                1   
1999-01-11           0                1                0                1   
1999-01-11           0                0                1                1   
1999-01-11           0                0      

In [71]:
# split the data

In [72]:
def train_validate_test_split(dfm):
    '''
    This function takes in a dataframe (df) and returns 3 dfs
    (train, validate, and test) split 20%, 24%, 56% respectively. 
    
    Also takes in a random seed for replicating results.  
    '''
    
    from sklearn.model_selection import train_test_split
     
    train_and_validate, test = train_test_split(
        dfm, test_size=0.2, random_state=123, stratify=dfm.player_1_wins
    )
    train, validate = train_test_split(
        train_and_validate,
        test_size=0.3,
        random_state=123,
        stratify=train_and_validate.player_1_wins,
    )
    return train, validate, test

In [73]:
train, validate, test = train_validate_test_split(dfm)

In [74]:
train.head()

draw_size              winner tourney_level  best_of  \
tourney_date                                                         
2015-10-11           56        Nick Kyrgios             M        3   
2017-05-22           32       Mischa Zverev             A        3   
2017-10-16           32  Stefanos Tsitsipas             A        3   
2006-10-09           32      Olivier Rochus             A        3   
2011-11-20           16       Roger Federer             F        3   

                           player_1             player_2  player_1_age  \
tourney_date                                                             
2015-10-11    Andreas Haider Maurer         Nick Kyrgios     28.558522   
2017-05-22            Kei Nishikori        Mischa Zverev     27.394935   
2017-10-16             Pablo Cuevas   Stefanos Tsitsipas     31.789185   
2006-10-09           Olivier Rochus  Paradorn Srichaphan     25.722108   
2011-11-20               Mardy Fish        Roger Federer     29.949350   

              player_2_age player_1_hand player_2_hand  player_1_ht  \
tourney_date                                                          
2015-10-11       20.459959             R             R        190.0   
2017-05-22       29.749487             R             L        178.0   
2017-10-16       19.178645             R             R        180.0   
2006-10-09       27.321013             R             R        168.0   
2011-11-20       30.286105             R             R        188.0   

              player_2_ht player_1_ioc player_2_ioc          player_1_name  \
tourney_date                                                                 
2015-10-11          193.0          AUT          AUS  Andreas Haider Maurer   
2017-05-22          190.0          JPN          GER          Kei Nishikori   
2017-10-16          193.0          URU          GRE           Pablo Cuevas   
2006-10-09          185.0          BEL          THA         Olivier Rochus   
2011-11-20          185.0          USA          SUI             Mardy Fish   

                    player_2_name  player_1_rank  player_2_rank  \
tourney_date                                                      
2015-10-11           Nick Kyrgios           60.0           32.0   
2017-05-22          Mischa Zverev            9.0           33.0   
2017-10-16     Stefanos Tsitsipas           33.0          122.0   
2006-10-09    Paradorn Srichaphan           30.0           41.0   
2011-11-20          Roger Federer            8.0            4.0   

              player_1_rank_points  player_2_rank_points  player_1_wins  \
tourney_date                                                              
2015-10-11                   785.0                1215.0          False   
2017-05-22                  3560.0                1165.0          False   
2017-10-16                  1325.0                 466.0          False   
2006-10-09                   990.0                 830.0           True   
2011-11-20                  2965.0                6670.0          False   

              round_ER  round_F  round_QF  round_R128  round_R16  round_R32  \
tourney_date                                                                  
2015-10-11           0        0         0           0          0          0   
2017-05-22           0        0         0           0          0          0   
2017-10-16           0        0         0           0          0          1   
2006-10-09           0        0         0           0          1          0   
2011-11-20           0        0         0           0          0          0   

              round_R64  round_RR  round_SF  player_1_hand_R  player_1_hand_L  \
tourney_date                                                                    
2015-10-11            1         0         0                1                0   
2017-05-22            0         0         1                1                0   
2017-10-16            0         0         0                1                0   
2006

Plan for explore: 

To identify features to use in modeling, separate the groups by wins / losses and then see if there was a significant difference between variables in that group. I'll take all of the matches that player1 wins and get the mean hight difference for player1 for that group then do the same for all of the matyches where player1 loses. If there is a significant difference then hight is likely a driver of winning and losing. 

reference: 
https://www.dataquest.io/blog/tutorial-add-column-pandas-dataframe-based-on-if-else-condition/

In [75]:
# seperate player 1 wins and losses into dfs

In [76]:
player_1_w = train[train['player_1_wins'] == True]
player_1_l = train[train['player_1_wins'] == False]


In [77]:
player_1_w.head(1)

draw_size          winner tourney_level  best_of  \
tourney_date                                                     
2006-10-09           32  Olivier Rochus             A        3   

                    player_1             player_2  player_1_age  player_2_age  \
tourney_date                                                                    
2006-10-09    Olivier Rochus  Paradorn Srichaphan     25.722108     27.321013   

             player_1_hand player_2_hand  player_1_ht  player_2_ht  \
tourney_date                                                         
2006-10-09               R             R        168.0        185.0   

             player_1_ioc player_2_ioc   player_1_name        player_2_name  \
tourney_date                                                                  
2006-10-09            BEL          THA  Olivier Rochus  Paradorn Srichaphan   

              player_1_rank  player_2_rank  player_1_rank_points  \
tourney_date                                                       
2006-10-09             30.0           41.0                 990.0   

              player_2_rank_points  player_1_wins  round_ER  round_F  \
tourney_date                                                           
2006-10-09                   830.0           True         0        0   

              round_QF  round_R128  round_R16  round_R32  round_R64  round_RR  \
tourney_date                                                                    
2006-10-09           0           0          1          0          0         0   

              round_SF  player_1_hand_R  player_1_hand_L  tourney_level_A  \
tourney_date                                                                
2006-10-09           0                1                0                1   

              tourney_level_D  tourney_level_F  tourney_level_G  \
tourney_date                                                      
2006-10-09                  0                0                0   

              tourney_level_M  surface_Carpet  surface_Clay  surface_Grass  \
tourney_date                                                                 
2006-10-09                  0               0             0              0   

              surface_Hard  winner_rank  loser_rank  ht_diff  age_diff  \
tourney_date                                                             
2006-10-09               1         30.0        41.0    -17.0 -1.598905   

              rank_diff  rank_points_diff  
tourney_date                               
2006-10-09        -11.0             160.0

In [78]:
player_1_w.shape

(9368, 47)

In [79]:
player_1_l.head(1)

draw_size        winner tourney_level  best_of  \
tourney_date                                                   
2015-10-11           56  Nick Kyrgios             M        3   

                           player_1      player_2  player_1_age  player_2_age  \
tourney_date                                                                    
2015-10-11    Andreas Haider Maurer  Nick Kyrgios     28.558522     20.459959   

             player_1_hand player_2_hand  player_1_ht  player_2_ht  \
tourney_date                                                         
2015-10-11               R             R        190.0        193.0   

             player_1_ioc player_2_ioc          player_1_name player_2_name  \
tourney_date                                                                  
2015-10-11            AUT          AUS  Andreas Haider Maurer  Nick Kyrgios   

              player_1_rank  player_2_rank  player_1_rank_points  \
tourney_date                                                       
2015-10-11             60.0           32.0                 785.0   

              player_2_rank_points  player_1_wins  round_ER  round_F  \
tourney_date                                                           
2015-10-11                  1215.0          False         0        0   

              round_QF  round_R128  round_R16  round_R32  round_R64  round_RR  \
tourney_date                                                                    
2015-10-11           0           0          0          0          1         0   

              round_SF  player_1_hand_R  player_1_hand_L  tourney_level_A  \
tourney_date                                                                
2015-10-11           0                1                0                0   

              tourney_level_D  tourney_level_F  tourney_level_G  \
tourney_date                                                      
2015-10-11                  0                0                0   

              tourney_level_M  surface_Carpet  surface_Clay  surface_Grass  \
tourney_date                                                                 
2015-10-11                  1               0             0              0   

              surface_Hard  winner_rank  loser_rank  ht_diff  age_diff  \
tourney_date                                                             
2015-10-11               1         32.0        60.0     -3.0  8.098563   

              rank_diff  rank_points_diff  
tourney_date                               
2015-10-11         28.0            -430.0

In [80]:
player_1_l.shape

(10011, 47)

Use .mean() to compare variables.

In [81]:
# Hight
print('Average hight, all matches: ' + str(train['player_1_ht'].mean()))
print('Average hight, player 1 wins: ' + str(player_1_w['player_1_ht'].mean()))
print('Average hight, player 1 loses: ' + str(player_1_l['player_1_ht'].mean()))
print('\n')

Average hight, all matches: 185.52144073481603
Average hight, player 1 wins: 185.77743381725023
Average hight, player 1 loses: 185.28188992108682




Doesn't look like there's a significant difference in the average hight amongst wins and losses. 

In [82]:
# Hight difference 
print('Average hight difference, all matches: ' + str(train['ht_diff'].mean()))
print('Average hight difference, player 1 wins: ' + str(player_1_w['ht_diff'].mean()))
print('Average hight difference, player 1 loses: ' + str(player_1_l['ht_diff'].mean()))
print('\n')

Average hight difference, all matches: -0.697972031580577
Average hight difference, player 1 wins: -0.15798462852263023
Average hight difference, player 1 loses: -1.203276395964439




In [83]:
# Age
print('Average age, all matches: ' + str(train['player_1_age'].mean()))
print('Average age, player 1 wins: ' + str(player_1_w['player_1_age'].mean()))
print('Average age, player 1 loses: ' + str(player_1_l['player_1_age'].mean()))
print('\n')

Average age, all matches: 26.42097101306274
Average age, player 1 wins: 26.325505558409912
Average age, player 1 loses: 26.510304783833845




Doesn't look like there's a significant difference in the average age amongst wins and losses.

In [84]:
# Age difference
print('Average age difference, all matches: ' + str(train['age_diff'].mean()))
print('Average age difference, player 1 wins: ' + str(player_1_w['age_diff'].mean()))
print('Average age difference, player 1 loses: ' + str(player_1_l['age_diff'].mean()))
print('\n')

Average age difference, all matches: -0.1596298822450129
Average age difference, player 1 wins: -0.3925796878822606
Average age difference, player 1 loses: 0.05835770932523186




In [85]:
# Rank
print('Average rank, all matches: ' + str(train['player_1_rank'].mean()))
print('Average rank, player 1 wins: ' + str(player_1_w['player_1_rank'].mean()))
print('Average rank, player 1 loses: ' + str(player_1_l['player_1_rank'].mean()))
print('\n')

Average rank, all matches: 60.62340678053563
Average rank, player 1 wins: 50.45420580700256
Average rank, player 1 loses: 70.1394466087304




It does look like there's a significant difference in the average rank amongst wins and losses.

In [86]:
# Rank points
print('Average rank points, all matches: ' + str(train['player_1_rank_points'].mean()))
print('Average rank points, player 1 wins: ' + str(player_1_w['player_1_rank_points'].mean()))
print('Average rank points, player 1 loses: ' + str(player_1_l['player_1_rank_points'].mean()))
print('\n')

Average rank points, all matches: 1321.1359203261263
Average rank points, player 1 wins: 1605.4534585824083
Average rank points, player 1 loses: 1055.0799120966935




In [87]:
print(train['player_1_ioc'].unique())

['AUT' 'JPN' 'URU' 'BEL' 'USA' 'SUI' 'ARG' 'ESP' 'TUN' 'KOR' 'AUS' 'CZE'
 'UZB' 'NED' 'ITA' 'RUS' 'GER' 'CRO' 'FRA' 'FIN' 'POL' 'GBR' 'BRA' 'CYP'
 'ECU' 'KAZ' 'MAR' 'MON' 'ROU' 'LUX' 'POR' 'UKR' 'RSA' 'SVK' 'THA' 'SRB'
 'COL' 'SWE' 'ARM' 'DEN' 'BIH' 'LAT' 'GEO' 'SLO' 'CHI' 'ISR' 'CRC' 'PER'
 'BUL' 'CAN' 'LTU' 'PHI' 'MDA' 'BLR' 'DOM' 'PAR' 'ZIM' 'TUR' 'HUN' 'TPE'
 'IND']


It does look like there's a significant difference in the average rank points amongst wins and losses.

 Because there's one continuous and one categorical variable that I want to test, I will use the (t-test).

In [88]:
# We want to be sure that the effect we see isn't explained by chance, so I will use a confidance interval of 95%
# The resulting alpha is .05.

null_hypothesis = "The average ranking when player 1 wins is the same as when player 1 loses."
alternative_hypothesis = "The average ranking when player 1 wins is not the same as when player 1 loses." 
alpha = 0.01

In [89]:
t, p = stats.ttest_1samp(train.player_1_wins, train.player_1_rank.mean())
print(f't = {t:.3f}')
print(f'p = {p:.3f}')

t = -16752.787
p = 0.000


In [90]:
if p < alpha:
    print("We reject the null hypothesis that", null_hypothesis)
    print("We move forward with the alternative hypothesis that", alternative_hypothesis)
else:
    print("We fail to reject the null")
    print("Our evidence does not support the claim that the average ranking when player 1 wins is not the same as when player 1 loses")


We reject the null hypothesis that The average ranking when player 1 wins is the same as when player 1 loses.
We move forward with the alternative hypothesis that The average ranking when player 1 wins is not the same as when player 1 loses.


In [91]:
# We want to be sure that the effect we see isn't explained by chance, so I will use a confidance interval of 95%
# The resulting alpha is .05.

null_hypothesis = "The average ranking points when player 1 wins is the same as when player 1 loses."
alternative_hypothesis = "The average ranking points when player 1 wins is not the same as when player 1 loses." 
alpha = 0.01

In [92]:
t, p = stats.ttest_1samp(train.player_1_wins, train.player_1_rank_points.mean())
print(f't = {t:.3f}')
print(f'p = {p:.3f}')

t = -367885.125
p = 0.000


In [93]:
if p < alpha:
    print("We reject the null hypothesis that", null_hypothesis)
    print("We move forward with the alternative hypothesis that", alternative_hypothesis)
else:
    print("We fail to reject the null")
    print("Our evidence does not support the claim that the average ranking points when player 1 wins is not the same as when player 1 loses")


We reject the null hypothesis that The average ranking points when player 1 wins is the same as when player 1 loses.
We move forward with the alternative hypothesis that The average ranking points when player 1 wins is not the same as when player 1 loses.


I want to look at hand and surface now to since those are both categorical i will be using the chi2 test.

In [94]:
# We want to be sure that the effect we see isn't explained by chance, so I will use a confidance interval of 95%
# The resulting alpha is .05.

In [95]:
null_hypothesis = "There is no dependence between right hand used and player 1 winning"
alternative_hypothesis = "There is a dependence between right hand used and player 1 winning"
alpha = 0.05

In [96]:
observed = pd.crosstab(train.player_1_hand_R, train.player_1_wins)

In [97]:
chi2, p, degf, expected = stats.chi2_contingency(observed)
p

0.00018815470417731938

In [98]:
if p < alpha:
    print("We reject the null that", null_hypothesis)
    print("We move forward with the alternative hypothesis that", alternative_hypothesis)
else:
    print("We fail to reject the null")
    print("Evidence does not support", alternative_hypothesis)

We reject the null that There is no dependence between right hand used and player 1 winning
We move forward with the alternative hypothesis that There is a dependence between right hand used and player 1 winning


In [99]:
def return_chi2(observed):
    
    #run the test
    chi2, p, degf, expected = stats.chi2_contingency(observed)
    
    #print the rest
    print('Observed')
    print('--------')
    print(observed.values)
    print('=====================')
    print('Expected')
    print('--------')
    print(expected.astype(int))
    print('=====================')
    print(f'Degrees of Freedom: {degf}')
    print('=====================')
    print('Chi^2 and P')
    print('-----------')
    print(f'chi^2 = {chi2:.4f}')
    print(f'p = {p:.4f}')

In [100]:
return_chi2(observed)

Observed
--------
[[1542 1265]
 [8469 8103]]
Expected
--------
[[1450 1356]
 [8560 8011]]
Degrees of Freedom: 1
Chi^2 and P
-----------
chi^2 = 13.9458
p = 0.0002


In [101]:
# We want to be sure that the effect we see isn't explained by chance, so I will use a confidance interval of 95%
# The resulting alpha is .05.

In [102]:
null_hypothesis = "There is no dependence between left hand used and player 1 winning"
alternative_hypothesis = "There is a dependence between left hand used and player 1 winning"
alpha = 0.05

In [103]:
observed = pd.crosstab(train.player_1_hand_L, train.player_1_wins)

In [104]:
chi2, p, degf, expected = stats.chi2_contingency(observed)
p

0.00018815470417731938

In [105]:
if p < alpha:
    print("We reject the null that", null_hypothesis)
    print("We move forward with the alternative hypothesis that", alternative_hypothesis)
else:
    print("We fail to reject the null")
    print("Evidence does not support", alternative_hypothesis)

We reject the null that There is no dependence between left hand used and player 1 winning
We move forward with the alternative hypothesis that There is a dependence between left hand used and player 1 winning


In [106]:
def return_chi2(observed):
    
    #run the test
    chi2, p, degf, expected = stats.chi2_contingency(observed)
    
    #print the rest
    print('Observed')
    print('--------')
    print(observed.values)
    print('=====================')
    print('Expected')
    print('--------')
    print(expected.astype(int))
    print('=====================')
    print(f'Degrees of Freedom: {degf}')
    print('=====================')
    print('Chi^2 and P')
    print('-----------')
    print(f'chi^2 = {chi2:.4f}')
    print(f'p = {p:.4f}')

In [107]:
return_chi2(observed)

Observed
--------
[[8469 8103]
 [1542 1265]]
Expected
--------
[[8560 8011]
 [1450 1356]]
Degrees of Freedom: 1
Chi^2 and P
-----------
chi^2 = 13.9458
p = 0.0002


In [108]:
# We want to be sure that the effect we see isn't explained by chance, so I will use a confidance interval of 95%
# The resulting alpha is .05.

In [109]:
null_hypothesis = "There is no dependence between carpet surface and player 1 winning"
alternative_hypothesis = "There is a dependence between carpet surface used and player 1 winning"
alpha = 0.05

In [110]:
observed = pd.crosstab(train.surface_Carpet, train.player_1_wins)

In [111]:
chi2, p, degf, expected = stats.chi2_contingency(observed)
p

0.2681556892572018

In [112]:
if p < alpha:
    print("We reject the null that", null_hypothesis)
    print("We move forward with the alternative hypothesis that", alternative_hypothesis)
else:
    print("We fail to reject the null")
    print("Evidence does not support", alternative_hypothesis)

We fail to reject the null
Evidence does not support There is a dependence between carpet surface used and player 1 winning


In [113]:
# We want to be sure that the effect we see isn't explained by chance, so I will use a confidance interval of 95%
# The resulting alpha is .05.

In [114]:
null_hypothesis = "There is no dependence between clay surface and player 1 winning"
alternative_hypothesis = "There is a dependence between clay surface used and player 1 winning"
alpha = 0.05

In [115]:
observed = pd.crosstab(train.surface_Clay, train.player_1_wins)

In [116]:
chi2, p, degf, expected = stats.chi2_contingency(observed)
p

0.029989066436845856

In [117]:
if p < alpha:
    print("We reject the null that", null_hypothesis)
    print("We move forward with the alternative hypothesis that", alternative_hypothesis)
else:
    print("We fail to reject the null")
    print("Evidence does not support", alternative_hypothesis)

We reject the null that There is no dependence between clay surface and player 1 winning
We move forward with the alternative hypothesis that There is a dependence between clay surface used and player 1 winning


In [118]:
# We want to be sure that the effect we see isn't explained by chance, so I will use a confidance interval of 95%
# The resulting alpha is .05.

In [119]:
null_hypothesis = "There is no dependence between grass surface and player 1 winning"
alternative_hypothesis = "There is a dependence between clay surface used and player 1 winning"
alpha = 0.05

In [120]:
observed = pd.crosstab(train.surface_Grass, train.player_1_wins)

In [121]:
chi2, p, degf, expected = stats.chi2_contingency(observed)
p

0.8081996624684368

In [122]:
if p < alpha:
    print("We reject the null that", null_hypothesis)
    print("We move forward with the alternative hypothesis that", alternative_hypothesis)
else:
    print("We fail to reject the null")
    print("Evidence does not support", alternative_hypothesis)

We fail to reject the null
Evidence does not support There is a dependence between clay surface used and player 1 winning


In [123]:
# We want to be sure that the effect we see isn't explained by chance, so I will use a confidance interval of 95%
# The resulting alpha is .05.

In [124]:
null_hypothesis = "There is no dependence between hard surface and player 1 winning"
alternative_hypothesis = "There is a dependence between hard surface used and player 1 winning"
alpha = 0.05

In [125]:
observed = pd.crosstab(train.surface_Hard, train.player_1_wins)

In [126]:
chi2, p, degf, expected = stats.chi2_contingency(observed)
p

0.1328077278331973

In [127]:
if p < alpha:
    print("We reject the null that", null_hypothesis)
    print("We move forward with the alternative hypothesis that", alternative_hypothesis)
else:
    print("We fail to reject the null")
    print("Evidence does not support", alternative_hypothesis)

We fail to reject the null
Evidence does not support There is a dependence between hard surface used and player 1 winning


In [128]:
# We want to be sure that the effect we see isn't explained by chance, so I will use a confidance interval of 95%
# The resulting alpha is .05.

In [129]:
null_hypothesis = "There is no dependence between ER round and player 1 winning"
alternative_hypothesis = "There is a dependence between hard surface used and player 1 winning"
alpha = 0.05

In [130]:
observed = pd.crosstab(train.round_ER, train.player_1_wins)

In [131]:
chi2, p, degf, expected = stats.chi2_contingency(observed)
p

0.686240419229087

In [132]:
if p < alpha:
    print("We reject the null that", null_hypothesis)
    print("We move forward with the alternative hypothesis that", alternative_hypothesis)
else:
    print("We fail to reject the null")
    print("Evidence does not support", alternative_hypothesis)

We fail to reject the null
Evidence does not support There is a dependence between hard surface used and player 1 winning


In [133]:
# We want to be sure that the effect we see isn't explained by chance, so I will use a confidance interval of 95%
# The resulting alpha is .05.

In [134]:
null_hypothesis = "There is no dependence between F round and player 1 winning"
alternative_hypothesis = "There is a dependence between F round and player 1 winning"
alpha = 0.05

In [135]:
observed = pd.crosstab(train.round_F, train.player_1_wins)

In [136]:
return_chi2(observed)

Observed
--------
[[9730 9095]
 [ 281  273]]
Expected
--------
[[9724 9100]
 [ 286  267]]
Degrees of Freedom: 1
Chi^2 and P
-----------
chi^2 = 0.1637
p = 0.6857


In [137]:
if p < alpha:
    print("We reject the null that", null_hypothesis)
    print("We move forward with the alternative hypothesis that", alternative_hypothesis)
else:
    print("We fail to reject the null")
    print("Evidence does not support", alternative_hypothesis)

We fail to reject the null
Evidence does not support There is a dependence between F round and player 1 winning


In [138]:
dfm = df[['draw_size',
       'tourney_level', 'best_of', 'player_1', 'player_2', 'player_1_age',
       'player_2_age', 'player_1_hand',
       'player_2_hand', 'player_1_ht', 'player_2_ht', 'player_1_ioc', 'player_2_ioc', 'player_1_name',
       'player_2_name', 'player_1_rank', 'player_2_rank',
       'player_1_rank_points', 'player_2_rank_points', 'player_1_wins', 'round_ER', 'round_F', 'round_QF', 'round_R128', 'round_R16', 'round_R32', 'round_R64', 'round_RR', 'round_SF', 'player_1_hand_R', 'player_1_hand_L', 
       'tourney_level_A', 'tourney_level_D', 'tourney_level_F', 'tourney_level_G', 'tourney_level_M', 'surface_Carpet', 'surface_Clay',
       'surface_Grass', 'surface_Hard']].copy(0)

In [139]:
# We want to be sure that the effect we see isn't explained by chance, so I will use a confidance interval of 95%
# The resulting alpha is .05.

In [140]:
null_hypothesis = "There is no dependence between round R32 and player 1 winning"
alternative_hypothesis = "There is a dependence between round R32 used and player 1 winning"
alpha = 0.05

In [141]:
observed = pd.crosstab(train.round_ER, train.player_1_wins)

In [142]:
return_chi2(observed)

Observed
--------
[[10006  9361]
 [    5     7]]
Expected
--------
[[10004  9362]
 [    6     5]]
Degrees of Freedom: 1
Chi^2 and P
-----------
chi^2 = 0.1632
p = 0.6862


In [143]:
if p < alpha:
    print("We reject the null that", null_hypothesis)
    print("We move forward with the alternative hypothesis that", alternative_hypothesis)
else:
    print("We fail to reject the null")
    print("Evidence does not support", alternative_hypothesis)

We fail to reject the null
Evidence does not support There is a dependence between round R32 used and player 1 winning


In [144]:
# We want to be sure that the effect we see isn't explained by chance, so I will use a confidance interval of 95%
# The resulting alpha is .05.

In [145]:
null_hypothesis = "There is no dependence between tourney level M and player 1 winning"
alternative_hypothesis = "There is a dependence between tourney level M used and player 1 winning"
alpha = 0.05

In [146]:
observed = pd.crosstab(train.tourney_level_M, train.player_1_wins)

In [147]:
return_chi2(observed)

Observed
--------
[[7733 7270]
 [2278 2098]]
Expected
--------
[[7750 7252]
 [2260 2115]]
Degrees of Freedom: 1
Chi^2 and P
-----------
chi^2 = 0.3377
p = 0.5612


In [148]:
if p < alpha:
    print("We reject the null that", null_hypothesis)
    print("We move forward with the alternative hypothesis that", alternative_hypothesis)
else:
    print("We fail to reject the null")
    print("Evidence does not support", alternative_hypothesis)

We fail to reject the null
Evidence does not support There is a dependence between tourney level M used and player 1 winning


In [149]:
# We want to be sure that the effect we see isn't explained by chance, so I will use a confidance interval of 95%
# The resulting alpha is .05.

In [150]:
null_hypothesis = "There is no dependence between tourney level G and player 1 winning"
alternative_hypothesis = "There is a dependence between tourney level G used and player 1 winning"
alpha = 0.05

In [151]:
observed = pd.crosstab(train.tourney_level_G, train.player_1_wins)

In [152]:
return_chi2(observed)

Observed
--------
[[7961 7477]
 [2050 1891]]
Expected
--------
[[7975 7462]
 [2035 1905]]
Degrees of Freedom: 1
Chi^2 and P
-----------
chi^2 = 0.2365
p = 0.6267


In [153]:
if p < alpha:
    print("We reject the null that", null_hypothesis)
    print("We move forward with the alternative hypothesis that", alternative_hypothesis)
else:
    print("We fail to reject the null")
    print("Evidence does not support", alternative_hypothesis)

We fail to reject the null
Evidence does not support There is a dependence between tourney level G used and player 1 winning


In [154]:
train['tourney_level'].value_counts() 

A    10802
M     4376
G     3941
F      156
D      104
Name: tourney_level, dtype: int64

In [155]:
df['round'].value_counts() 

R32     11100
R16      6664
R64      6048
R128     3735
QF       3612
SF       1898
F        1001
RR        714
ER         19
Name: round, dtype: int64

In [156]:
df['surface'].value_counts() 

Hard      18482
Clay      11663
Grass      3630
Carpet     1016
Name: surface, dtype: int64

In [157]:
# We want to be sure that the effect we see isn't explained by chance, so I will use a confidance interval of 95%
# The resulting alpha is .05.

In [158]:
null_hypothesis = "There is no dependence between age difference and player 1 winning"
alternative_hypothesis = "There is a dependence between age difference used and player 1 winning"
alpha = 0.05

In [159]:
observed = pd.crosstab(train.age_diff, train.player_1_wins)

In [160]:
return_chi2(observed)

Observed
--------
[[0 1]
 [1 0]
 [1 0]
 ...
 [0 1]
 [1 0]
 [0 1]]
Expected
--------
[[0 0]
 [0 0]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]
Degrees of Freedom: 11043
Chi^2 and P
-----------
chi^2 = 11487.6011
p = 0.0016


In [161]:
if p < alpha:
    print("We reject the null that", null_hypothesis)
    print("We move forward with the alternative hypothesis that", alternative_hypothesis)
else:
    print("We fail to reject the null")
    print("Evidence does not support", alternative_hypothesis)

We fail to reject the null
Evidence does not support There is a dependence between age difference used and player 1 winning


In [162]:
# We want to be sure that the effect we see isn't explained by chance, so I will use a confidance interval of 95%
# The resulting alpha is .05.

In [163]:
null_hypothesis = "There is no dependence between hight difference and player 1 winning"
alternative_hypothesis = "There is a dependence between hight difference used and player 1 winning"
alpha = 0.05

In [164]:
observed = pd.crosstab(train.ht_diff, train.player_1_wins)

In [165]:
return_chi2(observed)

Observed
--------
[[   1    0]
 [   1    1]
 [   8    1]
 [   6    5]
 [   6    2]
 [  27   15]
 [   0    1]
 [  14   12]
 [  28   19]
 [   2    0]
 [  83   47]
 [   2    0]
 [  48   48]
 [  84   56]
 [   1    4]
 [ 167  122]
 [  39   20]
 [  58   62]
 [ 255  204]
 [ 434  350]
 [  90   61]
 [ 111   79]
 [ 588  522]
 [  11    2]
 [ 525  515]
 [ 303  235]
 [  77   73]
 [ 875  832]
 [  11   11]
 [ 549  503]
 [ 522  442]
 [   1    2]
 [1054  997]
 [   4    2]
 [ 481  475]
 [ 630  589]
 [   1    3]
 [ 834  811]
 [  23   33]
 [ 263  292]
 [ 442  418]
 [   3    1]
 [ 405  438]
 [  93   78]
 [  76  100]
 [ 259  251]
 [   6    2]
 [ 186  190]
 [  52   56]
 [  12   16]
 [ 101  122]
 [   1    1]
 [  37   70]
 [  21   32]
 [   5    9]
 [  32   59]
 [  23   19]
 [  11   17]
 [  17   25]
 [   7    6]
 [   0    4]
 [   2    3]
 [   2    1]
 [   0    2]
 [   1    0]]
Expected
--------
[[   0    0]
 [   1    0]
 [   4    4]
 [   5    5]
 [   4    3]
 [  21   20]
 [   0    0]
 [  13   12]
 [  24   22]
 

In [166]:
if p < alpha:
    print("We reject the null that", null_hypothesis)
    print("We move forward with the alternative hypothesis that", alternative_hypothesis)
else:
    print("We fail to reject the null")
    print("Evidence does not support", alternative_hypothesis)

We fail to reject the null
Evidence does not support There is a dependence between hight difference used and player 1 winning


Features to move forward with: player_1_rank_points, player_1_hand_R, player_1_hand_L, surface_Clay. 

## Modeling

Split into x and y train for modeling:

In [167]:
train.columns

Index(['draw_size', 'winner', 'tourney_level', 'best_of', 'player_1',
       'player_2', 'player_1_age', 'player_2_age', 'player_1_hand',
       'player_2_hand', 'player_1_ht', 'player_2_ht', 'player_1_ioc',
       'player_2_ioc', 'player_1_name', 'player_2_name', 'player_1_rank',
       'player_2_rank', 'player_1_rank_points', 'player_2_rank_points',
       'player_1_wins', 'round_ER', 'round_F', 'round_QF', 'round_R128',
       'round_R16', 'round_R32', 'round_R64', 'round_RR', 'round_SF',
       'player_1_hand_R', 'player_1_hand_L', 'tourney_level_A',
       'tourney_level_D', 'tourney_level_F', 'tourney_level_G',
       'tourney_level_M', 'surface_Carpet', 'surface_Clay', 'surface_Grass',
       'surface_Hard', 'winner_rank', 'loser_rank', 'ht_diff', 'age_diff',
       'rank_diff', 'rank_points_diff'],
      dtype='object')

In [196]:
X_train = train.drop(columns=['draw_size', 'tourney_level', 'best_of', 'player_1', 'player_2',
       'player_1_age', 'player_2_age', 'player_1_hand', 'player_2_hand',
       'player_1_ht', 'player_2_ht', 'player_1_ioc', 'player_2_ioc',
       'player_1_name', 'player_2_name', 'player_1_rank', 'player_2_rank',
       'player_2_rank_points', 'player_1_wins',
       'round_ER', 'round_F', 'round_QF', 'round_R128', 'round_R16',
       'round_R32', 'round_R64', 'round_RR', 'round_SF', 'tourney_level_A', 'tourney_level_D',
       'tourney_level_F', 'tourney_level_G', 'tourney_level_M',
       'surface_Carpet', 'surface_Grass', 'surface_Hard',
       'ht_diff', 'age_diff', 'rank_diff', 'rank_points_diff', 'winner_rank', 'loser_rank', 'winner'])
y_train = train[['player_1_wins']]

X_validate = validate.drop(columns=['draw_size', 'tourney_level', 'best_of', 'player_1', 'player_2',
       'player_1_age', 'player_2_age', 'player_1_hand', 'player_2_hand',
       'player_1_ht', 'player_2_ht', 'player_1_ioc', 'player_2_ioc',
       'player_1_name', 'player_2_name', 'player_1_rank', 'player_2_rank',
       'player_2_rank_points', 'player_1_wins',
       'round_ER', 'round_F', 'round_QF', 'round_R128', 'round_R16',
       'round_R32', 'round_R64', 'round_RR', 'round_SF', 'tourney_level_A', 'tourney_level_D',
       'tourney_level_F', 'tourney_level_G', 'tourney_level_M',
       'surface_Carpet', 'surface_Grass', 'surface_Hard',
       'ht_diff', 'age_diff', 'rank_diff', 'rank_points_diff', 'winner_rank', 'loser_rank', 'winner'])
y_validate = validate[['player_1_wins']]

X_test = test.drop(columns=['draw_size', 'tourney_level', 'best_of', 'player_1', 'player_2',
       'player_1_age', 'player_2_age', 'player_1_hand', 'player_2_hand',
       'player_1_ht', 'player_2_ht', 'player_1_ioc', 'player_2_ioc',
       'player_1_name', 'player_2_name', 'player_1_rank', 'player_2_rank',
       'player_2_rank_points', 'player_1_wins',
       'round_ER', 'round_F', 'round_QF', 'round_R128', 'round_R16',
       'round_R32', 'round_R64', 'round_RR', 'round_SF', 'tourney_level_A', 'tourney_level_D',
       'tourney_level_F', 'tourney_level_G', 'tourney_level_M',
       'surface_Carpet', 'surface_Grass', 'surface_Hard',
       'ht_diff', 'age_diff', 'rank_diff', 'rank_points_diff', 'winner_rank', 'loser_rank', 'winner'])
y_test = test[[ 'player_1_wins']]

In [197]:
X_train.head()

player_1_rank_points  player_1_hand_R  player_1_hand_L  \
tourney_date                                                           
2015-10-11                   785.0                1                0   
2017-05-22                  3560.0                1                0   
2017-10-16                  1325.0                1                0   
2006-10-09                   990.0                1                0   
2011-11-20                  2965.0                1                0   

              surface_Clay  
tourney_date                
2015-10-11               0  
2017-05-22               1  
2017-10-16               0  
2006-10-09               0  
2011-11-20               0

In [198]:
y_train.head()

player_1_wins
tourney_date               
2015-10-11            False
2017-05-22            False
2017-10-16            False
2006-10-09             True
2011-11-20            False

### Baseline

Create a baseline: 

In [199]:
# df['winner_rank'] = np.where(df['winner'] == df['player_1_name'], df['player_1_rank_points'], df['player_2_rank_points'])
# df.head()

In [200]:
# df['loser_rank'] = np.where(df['winner'] == df['player_2_name'], df['player_1_rank_points'], df['player_2_rank_points'])
# df.head()

In [201]:
no_upset = train['winner_rank'] < train['loser_rank']


In [202]:
no_upset.head()

tourney_date
2015-10-11     True
2017-05-22    False
2017-10-16    False
2006-10-09     True
2011-11-20     True
dtype: bool

In [203]:
no_upset.value_counts(normalize=True)

True     0.644254
False    0.355746
dtype: float64

In [204]:
baseline = no_upset.mode()

In [205]:
baseline

0    True
dtype: bool

In [206]:
match_bsl_prediction = no_upset == 1

In [207]:
baseline_accuracy = match_bsl_prediction.mean()

In [208]:
# basline accuracy = 64%%
baseline_accuracy

0.6442540894783012

### Decision Tree

In [213]:
tree1_clf = DecisionTreeClassifier(max_depth=5, random_state=123)

In [214]:
tree1_clf = tree1_clf.fit(X_train, y_train)

In [215]:
#visualize the decision tree
import graphviz
from graphviz import Graph

dot_data = export_graphviz(tree1_clf, feature_names= X_train.columns, rounded=True, filled=True, out_file=None)
graph = graphviz.Source(dot_data) 

graph.render('tennis_decision_tree', view=True)

'tennis_decision_tree.pdf'

In [216]:
y_pred = tree1_clf.predict(X_train)
y_pred[0:5]

array([False,  True,  True, False,  True])

In [217]:
y_pred_proba = tree1_clf.predict_proba(X_train)
y_pred_proba[0:5]

array([[0.60361613, 0.39638387],
       [0.25854993, 0.74145007],
       [0.44119059, 0.55880941],
       [0.54463072, 0.45536928],
       [0.35826972, 0.64173028]])

In [218]:
# accuracy:
print('Accuracy of Decision Tree 1 classifier on training set: {:.2f}'
      .format(tree1_clf.score(X_train, y_train)))


Accuracy of Decision Tree 1 classifier on training set: 0.60


In [219]:
# confusion matrix
print(confusion_matrix(y_train, y_pred))


[[7582 2429]
 [5395 3973]]


In [220]:
pd.DataFrame(confusion_matrix(y_train, y_pred))

0     1
0  7582  2429
1  5395  3973

In [221]:
# classification report
print(classification_report(y_train, y_pred))

              precision    recall  f1-score   support

       False       0.58      0.76      0.66     10011
        True       0.62      0.42      0.50      9368

    accuracy                           0.60     19379
   macro avg       0.60      0.59      0.58     19379
weighted avg       0.60      0.60      0.58     19379



In [222]:
pd.DataFrame(classification_report(y_train, y_pred, output_dict=True))

False         True  accuracy     macro avg  weighted avg
precision      0.584264     0.620587  0.596264      0.602426      0.601823
recall         0.757367     0.424103  0.596264      0.590735      0.596264
f1-score       0.659649     0.503868  0.596264      0.581758      0.584343
support    10011.000000  9368.000000  0.596264  19379.000000  19379.000000

In [225]:
# max_depth = 16

# for i in range(1, max_depth):
#     # Create Model
#     depth = max_depth - i
#     n = i
#     forest = RandomForestClassifier(max_depth=depth, min_samples_leaf=n, random_state=123)

#     # Fit the model (on train and only train)
#     forest = forest.fit(X_train, y_train)

#     # Use the model
#     # We'll evaluate the model's performance on train, first
#     y_pred = forest.predict(X_train)

#     # Produce the classification report on the actual y values and this model's predicted y values
#     report = classification_report(y_train, y_pred, output_dict=True)
#     print(f"Tree with max depth of {i}")
#     print(pd.DataFrame(report))
#     print()

In [226]:
threshold = 0.10  #threshold set for amount of overfit that is tolerated

models = []
metrics = []

for i in range(1, 11):
    tree = DecisionTreeClassifier(max_depth=i, random_state=123)
    #^^^ creates the model
    
    tree = tree.fit(X_train, y_train)   #fit model to train data and only TRAIN data
    
    in_sample_accuracy = tree.score(X_train, y_train)
    out_sample_accuracy = tree.score(X_validate, y_validate)
    #^^^evaluates the models performance on train data first
    
    difference = in_sample_accuracy - out_sample_accuracy
    #^^calculates the difference in accuracy
    
    if difference > threshold:
        break
    #^^adds conditions to check the accuracy vs the threshold
    
    output = {
        'max_depth': i,
        'train_accuracy': in_sample_accuracy,
        'validate_accuracy': out_sample_accuracy,
        'difference': difference}
    #^^^formats the output for each models performance o train and validate
    
    metrics.append(output)
    
    models.append(output)
    
model_df = pd.DataFrame(metrics)
model_df["difference"] = model_df.train_accuracy - model_df.validate_accuracy


model_df.head()

max_depth  train_accuracy  validate_accuracy  difference
0          1        0.593065           0.590296    0.002768
1          2        0.593065           0.590296    0.002768
2          3        0.593065           0.590296    0.002768
3          4        0.593735           0.588490    0.005245
4          5        0.596264           0.588611    0.007653

The best model for decision tree is a model with the max depth of 3. 

### Random Forest

In [227]:
# create model
rf1_clf = RandomForestClassifier(max_depth=10, min_samples_leaf=1, random_state=123)

In [228]:
rf1_clf

RandomForestClassifier(max_depth=10, random_state=123)

In [229]:
rf1_clf.fit(X_train, y_train)

RandomForestClassifier(max_depth=10, random_state=123)

In [230]:
print(rf1_clf.feature_importances_)

[0.97868799 0.0033975  0.00405266 0.01386185]


In [231]:
y_pred = rf1_clf.predict(X_train)
y_pred

array([False,  True,  True, ..., False, False, False])

In [233]:
y_pred_proba = rf1_clf.predict_proba(X_train)

In [234]:
print('Accuracy of random forest classifier on training set: {:.2f}'
     .format(rf1_clf.score(X_train, y_train)))

Accuracy of random forest classifier on training set: 0.64


In [235]:
print(confusion_matrix(y_train, y_pred))

[[7612 2399]
 [4672 4696]]


In [236]:
pd.DataFrame(confusion_matrix(y_train, y_pred))

0     1
0  7612  2399
1  4672  4696

In [237]:
print(classification_report(y_train, y_pred))

              precision    recall  f1-score   support

       False       0.62      0.76      0.68     10011
        True       0.66      0.50      0.57      9368

    accuracy                           0.64     19379
   macro avg       0.64      0.63      0.63     19379
weighted avg       0.64      0.64      0.63     19379



In [238]:
class_report = classification_report(y_train, y_pred, output_dict=(True))
print("Tree1 depth")
pd.DataFrame(class_report)

Tree1 depth


False         True  accuracy     macro avg  weighted avg
precision      0.619668     0.661875   0.63512      0.640771      0.640071
recall         0.760364     0.501281   0.63512      0.630822      0.635120
f1-score       0.682844     0.570491   0.63512      0.626668      0.628531
support    10011.000000  9368.000000   0.63512  19379.000000  19379.000000

In [239]:
metrics = []
max_depth = 16

for i in range(1, max_depth):
    # Create model
    depth = max_depth - i
    n = i
    forest = RandomForestClassifier(max_depth=depth, min_samples_leaf=n, random_state=123)

    # Fit the model (on train and only train)
    forest = forest.fit(X_train, y_train)

    # Use the model
    # We'll evaluate the model's performance on train, first
    in_sample_accuracy = forest.score(X_train, y_train)
    
    out_of_sample_accuracy = forest.score(X_validate, y_validate)

    output = {
        "min_samples_per_leaf": n,
        "max_depth": depth,
        "train_accuracy": in_sample_accuracy,
        "validate_accuracy": out_of_sample_accuracy
    }
    
    metrics.append(output)
    
df = pd.DataFrame(metrics)
df["difference"] = df.train_accuracy - df.validate_accuracy
df

min_samples_per_leaf  max_depth  train_accuracy  validate_accuracy  \
0                      1         15        0.680530           0.554418   
1                      2         14        0.664379           0.562846   
2                      3         13        0.651427           0.572478   
3                      4         12        0.643738           0.575006   
4                      5         11        0.634553           0.575488   
5                      6         10        0.627587           0.578016   
6                      7          9        0.620775           0.580063   
7                      8          8        0.613344           0.581989   
8                      9          7        0.607307           0.585842   
9                     10          6        0.601837           0.586444   
10                    11          5        0.596625           0.588731   
11                    12          4        0.593581           0.590778   
12                    13          3        0.593168           0.590537   
13                    14          2        0.593065           0.590657   
14                    15          1        0.593065           0.590417   

    difference  
0     0.126112  
1     0.101533  
2     0.078949  
3     0.068732  
4     0.059065  
5     0.049571  
6     0.040712  
7     0.031355  
8     0.021465  
9     0.015394  
10    0.007894  
11    0.002803  
12    0.002631  
13    0.002407  
14    0.002648

The best random forest model is a model with  3 min_samples_per_leaf and a max_depth of	13. Train accuracy: 0.651427 Validate accuracy: 0.572478

### KNN

In [240]:
knn = KNeighborsClassifier(n_neighbors=5, weights='uniform')

In [241]:
knn.fit(X_train, y_train)

KNeighborsClassifier()

In [242]:
y_pred = knn.predict(X_train)

In [243]:
#y_pred
y_pred_proba = knn.predict_proba(X_train)

In [244]:
# Accuracy
print('Accuracy of KNN classifier on training set: {:.2f}'
     .format(knn.score(X_train, y_train)))

Accuracy of KNN classifier on training set: 0.65


In [245]:
# confusion matrix
print(confusion_matrix(y_train, y_pred))

[[6837 3174]
 [3554 5814]]


In [246]:
pd.DataFrame(confusion_matrix(y_train, y_pred))

0     1
0  6837  3174
1  3554  5814

In [247]:
# classification report
print(classification_report(y_train, y_pred))

              precision    recall  f1-score   support

       False       0.66      0.68      0.67     10011
        True       0.65      0.62      0.63      9368

    accuracy                           0.65     19379
   macro avg       0.65      0.65      0.65     19379
weighted avg       0.65      0.65      0.65     19379



In [248]:
pd.DataFrame(classification_report(y_train, y_pred, output_dict=True))

False         True  accuracy     macro avg  weighted avg
precision      0.657973     0.646862   0.65282      0.652418      0.652602
recall         0.682949     0.620623   0.65282      0.651786      0.652820
f1-score       0.670228     0.633471   0.65282      0.651850      0.652460
support    10011.000000  9368.000000   0.65282  19379.000000  19379.000000

In [249]:
metrics = []

# loop through different values of k
for k in range(1, 21):
            
    # define the thing
    knn = KNeighborsClassifier(n_neighbors=k)
    
    # fit the thing (remmeber only fit on training data)
    knn.fit(X_train, y_train)
    
    # use the thing (calculate accuracy)
    train_accuracy = knn.score(X_train, y_train)
    validate_accuracy = knn.score(X_validate, y_validate)
    
    output = {
        "k": k,
        "train_accuracy": train_accuracy,
        "validate_accuracy": validate_accuracy
    }
    
    metrics.append(output)


df = pd.DataFrame(metrics)
df["difference"] = df.train_accuracy - df.validate_accuracy
df

k  train_accuracy  validate_accuracy  difference
0    1        0.639713           0.520708    0.119005
1    2        0.625832           0.530941    0.094891
2    3        0.654213           0.530941    0.123272
3    4        0.639558           0.542620    0.096938
4    5        0.652820           0.540453    0.112367
5    6        0.644357           0.551288    0.093069
6    7        0.649105           0.544305    0.104799
7    8        0.642551           0.552733    0.089818
8    9        0.646576           0.548640    0.097937
9   10        0.637185           0.554178    0.083007
10  11        0.641364           0.560679    0.080685
11  12        0.635688           0.560679    0.075009
12  13        0.636359           0.556947    0.079412
13  14        0.633366           0.555261    0.078105
14  15        0.632540           0.557308    0.075232
15  16        0.627793           0.561522    0.066271
16  17        0.629599           0.562846    0.066753
17  18        0.624284           0.566699    0.057585
18  19        0.624439           0.564652    0.059787
19  20        0.621910           0.566699    0.055212

The best KNN model is a model with a k of 18. Train accuracy: 0.624284 Validate accuracy: 0.566699	

### Logistic Regression

In [250]:
# Create the model
logit = LogisticRegression(random_state=123)

In [251]:
# Fit the model with the features we've selected
logit.fit(X_train, y_train)

LogisticRegression(random_state=123)

In [252]:
# predict on the same features we fit on
y_pred = logit.predict(X_train)

In [254]:
print("Baseline is", round(baseline_accuracy, 2))
print('Accuracy of Logistic Regression classifier on training set: {:.2f}'
     .format(logit.score(X_train, y_train)))

Baseline is 0.64
Accuracy of Logistic Regression classifier on training set: 0.59


In [258]:
train_accuracy = logit.score(X_train, y_train)
validate_accuracy = logit.score(X_validate, y_validate)
train_accuracy, validate_accuracy

(0.5868207853862428, 0.5916205152901517)

In [256]:
difference = train_accuracy - validate_accuracy
difference

-0.004799729903908867

In [259]:
# All features and c=9
logit2 = LogisticRegression(C=9, random_state=123)


In [260]:
# Fit the model with the features we've selected
logit2.fit(X_train, y_train)

LogisticRegression(C=9, random_state=123)

In [261]:
# predict on the same features we fit on
y_pred = logit2.predict(X_train)

In [262]:
print("Baseline is", round(baseline_accuracy, 2))
print('Accuracy of Logistic Regression classifier on training set: {:.2f}'
     .format(logit2.score(X_train, y_train)))

Baseline is 0.64
Accuracy of Logistic Regression classifier on training set: 0.59


In [263]:
train_accuracy = logit2.score(X_train, y_train)
validate_accuracy = logit2.score(X_validate, y_validate)
train_accuracy, validate_accuracy

(0.5868207853862428, 0.5916205152901517)

In [264]:
difference = train_accuracy - validate_accuracy
difference

-0.004799729903908867

## Modeling Results: 

best model = logistic regression with an accuracy on validate date of 59%.